src/SSTClosGovernor.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "SSTClosGovernor.hpp"
+
+#include <unistd.h>
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+
+#include <vector>
+
+#include "SSTClosGovernorImp.hpp"
+#include "geopm/Exception.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "config.h"
+
+namespace geopm
+{
+    std::unique_ptr<SSTClosGovernor> SSTClosGovernor::make_unique(void)
+    {
+        return geopm::make_unique<SSTClosGovernorImp>();
+    }
+
+    std::shared_ptr<SSTClosGovernor> SSTClosGovernor::make_shared(void)
+    {
+        return std::make_shared<SSTClosGovernorImp>();
+    }
+
+    SSTClosGovernorImp::SSTClosGovernorImp()
+        : SSTClosGovernorImp(platform_io(), platform_topo())
+    {
+    }
+
+    SSTClosGovernorImp::SSTClosGovernorImp(PlatformIO &platform_io, const PlatformTopo &platform_topo)
+        : m_platform_io(platform_io)
+        , m_platform_topo(platform_topo)
+        , m_do_write_batch(false)
+        , m_is_enabled(true)
+        , m_clos_assoc_ctl_domain_type(m_platform_io.control_domain_type("SST::COREPRIORITY:ASSOCIATION"))
+        , m_num_clos_assoc_ctl_domain(m_platform_topo.num_domain(m_clos_assoc_ctl_domain_type))
+        , m_clos_config_ctl_domain_type(m_platform_io.control_domain_type("SST::COREPRIORITY:0:FREQUENCY_MIN"))
+        , m_num_clos_config_ctl_domain(m_platform_topo.num_domain(m_clos_config_ctl_domain_type))
+        , m_frequency_min(m_platform_io.read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0))
+        , m_frequency_sticker(m_platform_io.read_signal("CPU_FREQUENCY_STICKER", GEOPM_DOMAIN_BOARD, 0))
+        , m_frequency_max(m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0))
+        , m_last_clos(m_num_clos_assoc_ctl_domain, HIGH_PRIORITY)
+    {
+    }
+
+    bool SSTClosGovernor::is_supported(PlatformIO &platform_io)
+    {
+        bool result;
+
+        try {
+            result = platform_io.read_signal("SST::COREPRIORITY_SUPPORT:CAPABILITIES",
+                                             GEOPM_DOMAIN_BOARD, 0) > 0;
+            result &= platform_io.read_signal("SST::TURBOFREQ_SUPPORT:SUPPORTED",
+                                              GEOPM_DOMAIN_BOARD, 0) > 0;
+        }
+        catch (...) {
+            result = false;
+        }
+
+        return result;
+    }
+
+    void SSTClosGovernorImp::init_platform_io(void)
+    {
+        for (size_t ctl_idx = 0; ctl_idx != m_num_clos_assoc_ctl_domain; ++ctl_idx) {
+            m_clos_control_idx.push_back(m_platform_io.push_control(
+                "SST::COREPRIORITY:ASSOCIATION",
+                m_clos_assoc_ctl_domain_type,
+                static_cast<int>(ctl_idx)));
```

Comment:
Since that `size_t` is usually 64 bits wide, and `int` is usually 32 bits wide, is there ever a possibility where this conversion would overflow, for example when `ctl_idx > INT_MAX`? What is the maximum value that is stored in `ctl_idx`?

Reply:
`ctl_idx` has `m_num_clos_assoc_ctl_domain` as an upper-bound. `m_num_clos_assoc_ctl_domain` is initialized by a return value from `int PlatformTopo::num_domain(int)`